src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -403,30 +561,103 @@ public SymbolicMatch FindMatch(RegexRunnerMode mode, ReadOnlySpan<char> input, i
             }
             else
             {
-                Registers endRegisters = _pattern._info.ContainsLineAnchor ?
+                Registers endRegisters = _containsAnyAnchor ?
                     FindSubcaptures<FullInputReader>(input, matchStart, matchEnd, perThreadData) :
                     FindSubcaptures<NoZAnchorInputReader>(input, matchStart, matchEnd, perThreadData);
                 return new SymbolicMatch(matchStart, matchEnd - matchStart, endRegisters.CaptureStarts, endRegisters.CaptureEnds);
             }
         }
 
+        /// <summary>
+        /// This version of <see cref="FindEndPositionFallback"/> uses a different set of interfaces,
+        /// which don't check for many inner loop edge cases e.g. input end or '\n'.
+        /// All edge cases are handled before entering the loop.
+        /// </summary>
+        private int FindEndPositionOptimized<TOptimizedInputReader, TAcceleratedStateHandler, TOptimizedNullabilityHandler>(
+            ReadOnlySpan<char> input, int pos, long timeoutOccursAt, RegexRunnerMode mode, PerThreadData perThreadData)
+            where TOptimizedInputReader : struct, IOptimizedInputReader
+            where TAcceleratedStateHandler : struct, IAcceleratedStateHandler
+            where TOptimizedNullabilityHandler : struct, IOptimizedNullabilityHandler
+        {
+            // this initial state candidate is not really used in the common DFA case
+            // and could potentially be removed in the future
+            int initialStatePosCandidate = pos;
+            var currentState = new CurrentState(_dotstarredInitialStates[GetCharKind<FullInputReader>(input, pos - 1)]);
+            int endPos = NoMatchExists;
+            int lengthMinus1 = input.Length - 1;
+
+            while (true)
+            {
+                int innerLoopLength;
+                bool done;
+                if (currentState.NfaState is null)
+                {
+                    const int dfaCharsPerTimeoutCheck = 100000;
```

Comment:
```suggestion
                    const int DfaCharsPerTimeoutCheck = 100000;
```
Nit, consts start with capitals in repo style